rxandroidble/src/main/java/com/polidea/rxandroidble2/internal/util/LocationServicesOkObservableApi23Factory.java

```diff
@@ -0,0 +1,58 @@
+package com.polidea.rxandroidble2.internal.util;
+
+import android.annotation.TargetApi;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.location.LocationManager;
+import android.os.Build;
+
+import bleshadow.javax.inject.Inject;
+import io.reactivex.Observable;
+import io.reactivex.ObservableEmitter;
+import io.reactivex.ObservableOnSubscribe;
+import io.reactivex.functions.Cancellable;
+
+@TargetApi(Build.VERSION_CODES.KITKAT)
+public class LocationServicesOkObservableApi23Factory {
+    private final Context context;
+    private final LocationServicesStatus locationServicesStatus;
+
+    @Inject
+    LocationServicesOkObservableApi23Factory(
+            final Context context,
+            final LocationServicesStatus locationServicesStatus) {
+        this.context = context;
+        this.locationServicesStatus = locationServicesStatus;
+    }
+
+    public Observable<Boolean> get() {
+        return Observable.create(new ObservableOnSubscribe<Boolean>() {
+            @Override
+            public void subscribe(final ObservableEmitter<Boolean> emitter) throws Exception {
+                final boolean initialValue = locationServicesStatus.isLocationProviderOk();
+                final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
+                    @Override
+                    public void onReceive(Context context, Intent intent) {
+                        final boolean newValue = locationServicesStatus.isLocationProviderOk();
+                        emitter.onNext(newValue);
+                    }
+                };
+                emitter.setCancellable(new Cancellable() {
+                    @Override
+                    public void cancel() throws Exception {
+                        try {
+                            context.unregisterReceiver(broadcastReceiver);
+                        } catch (IllegalStateException ignored) {
```

Comment:
We may end up leaking a context if this will get called before the `subscribe` finishes

Reply:
Good catch @dariuszseweryn 👍 There may be a case like that. For example, considering the following execution flow:

* The program passes by line 51 at `if (!emitter.isDisposed()) {`. There's still at least a subscriber of this `Observable`, so the `context.registerReceiver()` is about to be executed.
* Then, somehow no subscriber subscribes to the `Observable` anymore. The code in `setCancellable()` gets invoked.
* Then `context.registerReceiver()` gets executed. Thus the leak has just happened.

I'm thinking about doing another check `if (!emitter.isDisposed())` right after `context.registerReceiver()` but it looks kinda weird. I never wrote such code before. Wdyt?

Reply:
Or another idea is to move the `setCancellable()` into the `isDispose()` check. For example:

```java
if (!emitter.isDisposed()) {
    context.registerReceiver(broadcastReceiver, new IntentFilter(LocationManager.MODE_CHANGED_ACTION));
    emitter.onNext(initialValue);
    emitter.setCancellable(new Cancellable() {
        @Override
        public void cancel() throws Exception {
            try {
                context.unregisterReceiver(broadcastReceiver);
            } catch (IllegalStateException ignored) {
            }
        }
    });
}
```

Wdyt?

Reply:
I've changed the impl into:

```java
    public Observable<Boolean> get() {
        return Observable.create(new ObservableOnSubscribe<Boolean>() {
            @Override
            public void subscribe(final ObservableEmitter<Boolean> emitter) throws Exception {
                final boolean initialValue = locationServicesStatus.isLocationProviderOk();
                final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
                    @Override
                    public void onReceive(Context context, Intent intent) {
                        final boolean newValue = locationServicesStatus.isLocationProviderOk();
                        emitter.onNext(newValue);
                    }
                };
                emitter.onNext(initialValue);
                context.registerReceiver(broadcastReceiver, new IntentFilter(LocationManager.MODE_CHANGED_ACTION));
                emitter.setCancellable(new Cancellable() {
                    @Override
                    public void cancel() throws Exception {
                        try {
                            context.unregisterReceiver(broadcastReceiver);
                        } catch (IllegalStateException ignored) {
                        }
                    }
                });
            }
        }).distinctUntilChanged();
    }
```

I think this impl is able to prevent the above leak from happening.

Reply:
Yup, but we should get rid of the `try/catch` in the `Cancellable` as it could only hide potential issues

Reply:
@dariuszseweryn Good call 👍 Done.